### PR TITLE
Simplify SpreadMinimizingTokenGenerator for easier reuse in partitions ring.

### PR DESCRIPTION
**Pull request by pstibrany**
**What this PR does**:

This PR simplifies internals of  `SpreadMinimizingTokenGenerator`: it removes logger, slice with zones, parsing of instance when computing previous instance.

These changes make it easier to reuse this token generator in partitions ring code (https://github.com/grafana/dskit/compare/partitions-ring branch).

**Checklist**
- [x] Tests updated
- [na] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: ring/spread_minimizing_token_generator.go [hunk 1]
```diff
@@ -193,7 +169,11 @@ func (t *SpreadMinimizingTokenGenerator) GenerateTokens(requestedTokensCount int
 		used[v] = true
 	}
 
-	allTokens := t.generateAllTokens()
+	allTokens, err := t.generateAllTokens()
+	if err != nil {
+		// we were unable to generate required tokens, so we panic.
+		panic(err)
```
Comment (pstibrany) [inline ring/spread_minimizing_token_generator.go:175]: This `panic` is not new, it was just moved from `generateTokensByInstanceID` to this method.

File: ring/spread_minimizing_token_generator.go [hunk 2]
```diff
@@ -331,21 +310,18 @@ func (t *SpreadMinimizingTokenGenerator) generateTokensByInstanceID() map[int]To
...
 }
 
 func (t *SpreadMinimizingTokenGenerator) CanJoin(instances map[string]InstanceDesc) error {
 	if !t.canJoinEnabled {
 		return nil
 	}
 
-	prevInstance, err := previousInstance(t.instance)
-	if err != nil {
-		if errors.Is(err, errorNoPreviousInstance) {
```
Comment (duricanikolic) [inline ring/spread_minimizing_token_generator.go:344]: Could you please check whether we still need `errorNoPreviousInstance`? I think that at this point it should be removed.

**Review by duricanikolic — APPROVED** [review]
LGTM
Before merging could you please check the comment I wrote?

File: ring/spread_minimizing_token_generator.go (hunk 2 above)
Comment (pstibrany) [reply ring/spread_minimizing_token_generator.go:344]: This PR removed `errorNoPreviousInstance` variable.

**Pull request merged**